Seemingly different behavior of System.Runtime.Loader on Windows and Linux while trying to load analyzer assemblies in Roslyn

Repro: If you clone my branch of Roslyn at https://github.com/agocke/roslyn/tree/CoreClrAnalyzerLoadFailure and run "./cibuild" and record all the output you will see a bunch of warnings about being unable to load analyzer assemblies. I've attempted to add print statements to instrument the compiler to give more information about the load failure.

Short description of the problem, as far as I see it:

On Linux, and only on Linux, when Roslyn attempts to load any analyzer assembly using the CoreClrAnalyzerLoader (built off of AssemblyLoadContext) the CLR produces an exception while attempting to load the Microsoft.CodeAnalysis DLL, e.g.

```  
CSC : warning CS8032: An instance of analyzer Roslyn.Diagnostics.Analyzers.CSharp.CSharpDoNotCallContinueWithWithoutTaskSchedulerAnalyzer cannot be created from /home/andy/.nuget/packages/Microsoft.Net.RoslynDiagnostics/1.1.1-beta1-20150818-01/tools/Roslyn.Diagnostics.Analyzers.CSharp.dll : Could not load file or assembly 'Microsoft.CodeAnalysis, Version=1.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. Could not find or load a specific file. (Exception from HRESULT: 0x80131621). [/home/andy/code/roslyn/src/Compilers/Core/Portable/CodeAnalysis.csproj]
```

From what I can tell, this HRESULT indicates that we are attempting to load Microsoft.CodeAnalysis, even though it is already loaded. This seems like a sensible limitation, but when I've instrumented the Loader type to print all attempted loads to stderr, I see no information that Microsoft.CodeAnalysis was previously loaded nor had been requested to be loaded. It's possible that it's being loaded as a dependency of Csc before we have a change to set the default loader, but we're currently setting the default loader in Main (https://github.com/agocke/roslyn/blob/CoreClrAnalyzerLoadFailure/src/Compilers/CSharp/CscCore/Program.cs).

I'm not exactly sure if this is a bug, or by design, but it seems strange that it's different on Linux and Windows. If it is by design I would be grateful for some guidance on what the 'safe' loader code would look like, here.

--- a/src/binder/assemblybinder.cpp
+++ b/src/binder/assemblybinder.cpp
@@ -1228,17 +1228,9 @@ namespace BINDER_SPACE
 #ifdef FEATURE_LEGACYNETCF
         fWindowsPhone7 = RuntimeIsLegacyNetCF(pApplicationContext->GetAppDomainId()) == TRUE;
 #endif
-        //
-        // Windows Phone 7 Quirk:
-        //
-        // NetCF allows partial binds to platform assemblies.  If we are running a
-        // Mango application, skip the PKT check if no Ref version is provided,
-        // since there are apps in the Marketplace that do Assembly.Load("System")
-        //
+
         if (!tpaListAssembly || (fWindowsPhone7 && tpaListAssembly))
         {
-            dwIncludeFlags |= AssemblyName::EXCLUDE_PUBLIC_KEY_TOKEN_IF_MISSING;
-
             //
             // On Windows Phone 7, exclude culture comparisons when requesting an uncultured
             // assembly for app compat reasons (there are main app assemblies with spurious cultures)

--- a/src/binder/assemblyname.cpp
+++ b/src/binder/assemblyname.cpp
@@ -568,10 +568,10 @@ Exit:
             {
                 fEquals = EqualsCaseInsensitive(GetNormalizedCulture(), pAssemblyName->GetNormalizedCulture());
             }
-            if (fEquals && ((dwIncludeFlags & EXCLUDE_PUBLIC_KEY_TOKEN_IF_MISSING) == 0 ||
-                (pAssemblyName->Have(AssemblyIdentity::IDENTITY_FLAG_PUBLIC_KEY_TOKEN))))
+
+            if (fEquals && (dwIncludeFlags & INCLUDE_PUBLIC_KEY_TOKEN) != 0)
             {
-                    fEquals = (GetPublicKeyTokenBLOB().Equals(pAssemblyName->GetPublicKeyTokenBLOB()));
+                fEquals = (GetPublicKeyTokenBLOB().Equals(pAssemblyName->GetPublicKeyTokenBLOB()));
             }
             
             if (fEquals && ((dwIncludeFlags & INCLUDE_ARCHITECTURE) != 0))

--- a/src/binder/inc/assemblyname.hpp
+++ b/src/binder/inc/assemblyname.hpp
@@ -31,7 +31,7 @@ namespace BINDER_SPACE
             INCLUDE_ARCHITECTURE                = 0x02,
             INCLUDE_RETARGETABLE                = 0x04, 
             INCLUDE_CONTENT_TYPE                = 0x08,
-            EXCLUDE_PUBLIC_KEY_TOKEN_IF_MISSING = 0x10,
+            INCLUDE_PUBLIC_KEY_TOKEN            = 0x10,
             EXCLUDE_CULTURE                     = 0x20
         } INCLUDE_FLAGS;
 

